source/MaterialXCore/Definition.h

```diff
@@ -378,6 +390,92 @@ class Member : public TypedElement
     static const string CATEGORY;
 };
 
+/// @class UnitDef
+/// A unit definition within a UnitTypeDef.
+class UnitDef : public Element
+{
+  public:
+      UnitDef(ElementPtr parent, const string& name) :
+          Element(parent, CATEGORY, name)
+      {
+      }
+      virtual ~UnitDef() { }
+
+  public:
+    static const string CATEGORY;
+};
+
+/// @class UnitTypeDef
+/// A unit type definition element within a Document.
+class UnitTypeDef : public Element
+{
+  public:
+    UnitTypeDef(ElementPtr parent, const string& name) :
+        Element(parent, CATEGORY, name)
+    {
+    }
+    virtual ~UnitTypeDef() { }
+
+    /// @name Semantic
+    /// @{
+
+    /// Set the defaut string of the UnitTypeDef.
+    void setDefault(const string& value)
+    {
+        setAttribute(DEFAULT_ATTRIBUTE, value);
+    }
+
+    /// Return true if the given TypeDef has a default string.
+    bool hasDefault() const
+    {
+        return hasAttribute(DEFAULT_ATTRIBUTE);
+    }
+
+    /// Return the default string of the UnitTypeDef.
+    const string& getDefault() const
+    {
+        return getAttribute(DEFAULT_ATTRIBUTE);
+    }
+
+    /// @}
+    /// @name UnitDef Elements
+    /// @{
+
+    /// Add a UnitDef to the UnitTypeDef.
+    /// @param name The name of the new Member.
+    ///     If no name is specified, then a unique name will automatically be
+    ///     generated.
+    /// @return A shared pointer to the new UnitDef.
+    UnitDefPtr addUnitDef(const string& name = EMPTY_STRING)
```

Comment:
Should we allow empty names here? Isn't the name what identifies a unit, if automatically generated I don't see how the unit could be used.

Reply:
Did this via cut and paste. Yes, it probably doesn't make sense to make up unit names. I'll fix this.

Reply:
An exception is thrown now if the name is empty. The add child logic will already throw an exception if it's a duplicate.